opal/util/json/opal_json.c

```diff
@@ -0,0 +1,263 @@
+    if (json->value && NULL == json->value->parent) {
+        json_value_free(json->value);
+    }
+
+    free(json);
+}
+
+static inline int opal_json_internal_translate_type(json_type type, opal_json_type *out)
+{
+    int ret = OPAL_SUCCESS;
+    switch (type) {
+    case json_null:
+        *out = OPAL_JSON_NULL;
+        break;
+    case json_array:
+        *out = OPAL_JSON_ARRAY;
+        break;
+    case json_boolean:
+        *out = OPAL_JSON_BOOL;
+        break;
+    case json_integer:
+        *out = OPAL_JSON_INTEGER;
+        break;
+    case json_double:
+        *out = OPAL_JSON_DOUBLE;
+        break;
+    case json_string:
+        *out = OPAL_JSON_STRING;
+        break;
+    case json_object:
+        *out = OPAL_JSON_OBJECT;
+        break;
+    default:
+        opal_output_verbose(1, 0, "Invalid JSON value type %d\n", type);
+        *out = OPAL_JSON_TYPE_COUNT;
+        ret = OPAL_ERROR;
+    }
+
+    return ret;
+}
+
+static int opal_json_internal_new(const json_value *in, opal_json_internal_t **out)
+{
+    *out = malloc(sizeof(opal_json_internal_t));
+    if (NULL == *out) {
+        return OPAL_ERROR;
+    }
+
+    (*out)->value = (json_value *) in;
+    return opal_json_internal_translate_type(in->type, &(*out)->parent.type);
+}
+
+int opal_json_load(const char *str, const size_t len, opal_json_t **json)
+{
+    opal_json_internal_t *out = NULL;
+    int ret = OPAL_SUCCESS;
+
+    json_value *value = json_parse(str, len);
+    if (!value) {
+        opal_output_verbose(1, 0, "Unable to parse JSON string\n");
+        ret = OPAL_ERROR;
+        goto out;
+    }
+
+    ret = opal_json_internal_new(value, &out);
+
+out:
+    if (OPAL_SUCCESS == ret) {
+        *json = (opal_json_t *) out;
+    } else if (out) {
+        opal_json_internal_free(out);
+    }
+
+    return ret;
+}
+
+int opal_json_load_file(const char *filename, opal_json_t **json)
+{
+    FILE *fp = NULL;
+    size_t file_size;
+    char *file_contents = NULL;
+    int ret = OPAL_SUCCESS;
+
+    fp = fopen(filename, "r");
+    if (fp == NULL) {
+        opal_output_verbose(1, 0, "Unable to open %s\n", filename);
+        ret = OPAL_ERROR;
+        goto out;
+    }
+
+    (void) fseek(fp, 0L, SEEK_END);
+    file_size = ftell(fp);
+    rewind(fp);
+
+    file_contents = (char *) malloc(file_size);
+    if (!file_contents) {
+        opal_output_verbose(1, 0, "Unable to allocate memory\n");
+        ret = OPAL_ERROR;
+        goto out;
+    }
+
+    if (file_size > fread(file_contents, 1, file_size, fp)) {
+        opal_output_verbose(1, 0, "Unable to read content of %s\n", filename);
+        ret = OPAL_ERROR;
+        goto out;
+    }
+
+    ret = opal_json_load(file_contents, file_size, json);
+
+out:
+    if (fp) {
+        fclose(fp);
+    }
+    if (file_contents) {
+        free(file_contents);
+    }
+
+    return ret;
+}
+
+int opal_json_get_key(const opal_json_t *json, const char *key, opal_json_t **out)
```

Comment:
shouldn't the returned json object be const?

Reply:
I updated the signatures with const. The only uncertain one is `opal_json_free(opal_json_t *json)` which I don't think should take const input, since the function actually changes the object. This might become cumbersome over time, since the application needs to force cast the time everywhere, e.g.
```
const opal_json_t *json = NULL;

ret = opal_json_load_file("my_file.json", &json);
if (OPAL_SUCCESS != ret) {
     goto out;
}
opal_json_free((opal_json_t *) json);
```

Reply:
if you change free to take a `**` then this goes away right ? And the signature is in sync with opal_json_load_file

Reply:
You are right!

But now I'm confused about `const **` semantics. I thought both `const opal_json_t **json` and `const opal_json_t *json` imply that the `opal_json_t` object should not be modified, which contradicts the intention of `opal_json_free`. But obviously the compiler didn't think so.

Reply:
In the `const **` case the object itself is const, but you are modifying a pointer to such an object.